src/core/control/xojfile/XmlParser.cpp

```diff
@@ -0,0 +1,808 @@
+    while (xmlTextReaderMoveToNextAttribute(this->reader)) {
+        attributeMap[currentName()] = reinterpret_cast<const char*>(xmlTextReaderConstValue(this->reader));
+    }
+
+    DEBUG_PARSER(debugPrintAttributes(attributeMap));
+
+    return attributeMap;
+}
+
+
+auto XmlParser::openTag() -> TagType {
+    const TagType type = currentTagType();
+    // Add a level to the hierarchy only if the element isn't "empty" (which
+    // means there is no closing element)
+    if (!xmlTextReaderIsEmptyElement(this->reader)) {
+        this->hierarchy.push(type);
+    }
+    return type;
+}
+
+void XmlParser::closeTag(TagType type) {
+    // Check that the document structure is not messed up
+    if (this->hierarchy.empty()) {
+        throw std::runtime_error(
+                FS(_F("Error parsing XML file: found closing tag \"{1}\" at document root") % TAG_NAMES[type]));
+    }
+    if (this->hierarchy.top() != type) {
+        throw std::runtime_error(
+                FS(_F("Error parsing XML file: closing tag \"{1}\" does not correspond to last open element \"{2}\"") %
+                   TAG_NAMES[type] % TAG_NAMES[this->hierarchy.top()]));
+    }
+
+    // Go up one level in the hierarchy
+    this->hierarchy.pop();
+}
+
+auto XmlParser::tagNameToType(const std::string& name) const -> TagType {
+    static const std::unordered_map<std::string, TagType> typeMap = {{"xournal", TagType::XOURNAL},
+                                                                     {"MrWriter", TagType::MRWRITER},
+                                                                     {"title", TagType::TITLE},
+                                                                     {"preview", TagType::PREVIEW},
+                                                                     {"page", TagType::PAGE},
+                                                                     {"audio", TagType::AUDIO},
+                                                                     {"background", TagType::BACKGROUND},
+                                                                     {"layer", TagType::LAYER},
+                                                                     {"timestamp", TagType::TIMESTAMP},
+                                                                     {"stroke", TagType::STROKE},
+                                                                     {"text", TagType::TEXT},
+                                                                     {"image", TagType::IMAGE},
+                                                                     {"teximage", TagType::TEXIMAGE},
+                                                                     {"attachment", TagType::ATTACHMENT}};
+
+    auto it = typeMap.find(name);
+    if (it == typeMap.end()) {
+        g_warning("XML parser: unknown tag name \"%s\"", name.c_str());
+        return TagType::UNKNOWN;
+    } else {
+        return it->second;
+    }
+}
```

Comment:
I actually disagree to use a std::unordered map:
The function is called during a loop and the unordered map is known to be slow, puts values into buckets, which cause cache misses and memory fragmentation and hashing of strings is also slow. 
std::map would perform better, but it has the same memory issues (no flat map, memory fragmentation, cache misses).
I would use a `constexpr` array with sorted values:

```suggestion
auto XmlParser::tagNameToType(std::string_view name) -> TagType {
    using namespace std::literals;
    using Elem = std::pair<std::string_view, TagType>;
    // Todo(c++26): std::ranges::sort
    constexpr std::array typeMap{
        Elem{"MrWriter"sv, TagType::MRWRITER},
        Elem{"attachment"sv, TagType::ATTACHMENT},
        Elem{"audio"sv, TagType::AUDIO},
        Elem{"background"sv, TagType::BACKGROUND},
        Elem{"image"sv, TagType::IMAGE},
        Elem{"layer"sv, TagType::LAYER},
        Elem{"page"sv, TagType::PAGE},
        Elem{"preview"sv, TagType::PREVIEW},
        Elem{"stroke"sv, TagType::STROKE},
        Elem{"teximage"sv, TagType::TEXIMAGE},
        Elem{"text"sv, TagType::TEXT},
        Elem{"timestamp"sv, TagType::TIMESTAMP},
        Elem{"title"sv, TagType::TITLE},
        Elem{"xournal"sv, TagType::XOURNAL}};
    constexpr auto less_same = [](Elem const& lhs, Elem const& rhs) -> bool {
        return lhs.first.compare(rhs.first) < 0;
    };
    constexpr auto less_search = [](Elem const& lhs, std::string_view rhs) {
        return lhs.first < rhs;
    };
    // Todo (cpp20): static_assert:
    xoy_assert(std::is_sorted(begin(typeMap), end(typeMap), less_same));
    auto iter = std::lower_bound(begin(typeMap), end(typeMap), name, less_search);
    if (!(iter == end(typeMap)) && !(name < iter->first)) {
        return iter->second;
    }
    printf("XML parser: unknown tag name \"%*s\"", int(unsigned(name.size())),
              name.data());
    return TagType::UNKNOWN;
}

```
Unfortunately, I couldn't run a benchmark on this, since quick bench was down. But the code it produces is far smaller than the code of if else if else blocks and the current solution.
https://godbolt.org/z/bK3zT8Gjz

Reply:
Actually, this is faster:

```suggestion
auto tagNameToType2(std::string_view name) -> TagType {
    using namespace std::literals;

    if("MrWriter"sv == name) return TagType::MRWRITER;
    if("attachment"sv == name) return TagType::ATTACHMENT;
    if("audio"sv == name) return TagType::AUDIO;
    if("background"sv == name) return TagType::BACKGROUND;
    if("image"sv == name) return TagType::IMAGE;
    if("layer"sv == name) return TagType::LAYER;
    if("page"sv == name) return TagType::PAGE;
    if("preview"sv == name) return TagType::PREVIEW;
    if("stroke"sv == name) return TagType::STROKE;
    if("teximage"sv == name) return TagType::TEXIMAGE;
    if("text"sv == name) return TagType::TEXT;
    if("timestamp"sv == name) return TagType::TIMESTAMP;
    if("title"sv == name) return TagType::TITLE;
    if("xournal"sv == name) return TagType::XOURNAL;
    return TagType::UNKNOWN;
}
```

Reply:
![grafik](https://github.com/xournalpp/xournalpp/assets/2228627/2983a86b-5fe4-4399-8b15-e7ab7058462a)

Reply:
Actually the compiler does some nifty magic here, when I use a `constexpr std::array` vs a `constexpr static std::array` the if else chain performs just like the looped array. It also buffles me, that the unordered map is that fast, but only for constexpr std::string_views/cstrings, but not for std::string.

Reply:
I don't have the necessary education to understand what is happening in there, but if I use a if return, wouldn't it make sense to put more frequently used tags at the top (like Stroke)? What's the idea behind having it sorted alphabetically?

Reply:
> I don't have the necessary education to understand what is happening in there, but if I use a if return, wouldn't it make sense to put more frequently used tags at the top (like Stroke)? What's the idea behind having it sorted alphabetically?

Was just an experiment, for the sorted array, and I copied the code.
Actually it does not make any difference, whether it's sorted or if the most used tag is on top. The compiler resorts the whole if_else chain to form a binary tree.